pkg/import/importer.go

```diff
@@ -0,0 +1,467 @@
+package _import
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"path"
+
+	"github.com/ghodss/yaml"
+	"github.com/google/go-containerregistry/pkg/authn"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+	"github.com/pivotal/kpack/pkg/client/clientset/versioned"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/pivotal/build-service-cli/pkg/clusterstack"
+	"github.com/pivotal/build-service-cli/pkg/clusterstore"
+	"github.com/pivotal/build-service-cli/pkg/commands"
+	"github.com/pivotal/build-service-cli/pkg/config"
+	"github.com/pivotal/build-service-cli/pkg/k8s"
+)
+
+type ImageRelocator interface {
+	Relocate(keychain authn.Keychain, src v1.Image, destination string) (string, error)
+}
+
+type ImageFetcher interface {
+	Fetch(keychain authn.Keychain, image string) (v1.Image, error)
+}
+
+type TimestampProvider interface {
+	GetTimestamp() string
+}
+
+type Printer interface {
+	Printlnf(format string, args ...interface{}) error
+	PrintStatus(format string, args ...interface{}) error
+}
+
+type Importer struct {
+	client              versioned.Interface
+	k8sClient           kubernetes.Interface
+	printer             Printer
+	imageRelocator      ImageRelocator
+	imageFetcher        ImageFetcher
+	waiter              commands.ResourceWaiter
+	clusterStoreFactory *clusterstore.Factory
+	clusterStackFactory *clusterstack.Factory
+	timestampProvider   TimestampProvider
+	objs                []runtime.Object
+}
+
+type relocatedDescriptor struct {
+	lifecycle       *corev1.ConfigMap
+	clusterStores   []*v1alpha1.ClusterStore
+	clusterStacks   []*v1alpha1.ClusterStack
+	clusterBuilders []*v1alpha1.ClusterBuilder
+}
+
+func NewImporter(printer Printer, k8sClient kubernetes.Interface, client versioned.Interface, fetcher ImageFetcher, relocator ImageRelocator, waiter commands.ResourceWaiter, timestampProvider TimestampProvider) *Importer {
+	return &Importer{
+		imageRelocator:      relocator,
+		client:              client,
+		k8sClient:           k8sClient,
+		printer:             printer,
+		waiter:              waiter,
+		imageFetcher:        fetcher,
+		timestampProvider:   timestampProvider,
+		clusterStackFactory: clusterstack.NewFactory(printer, relocator, fetcher),
+		clusterStoreFactory: clusterstore.NewFactory(printer, relocator, fetcher),
+	}
+}
+
+func (i *Importer) ReadDescriptor(descriptorReader io.Reader) (DependencyDescriptor, error) {
+	return readDescriptor(descriptorReader)
```

Comment:
Should we just inline this method?

Reply:
ya we can, the ImportDescriptor method can just call the ReadDescriptor Method instead

Reply:
i think this is leftover when I had the dry run importer struct that shared this logic